### PR TITLE
Implement Sets in the IR

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: src/main/scala/is/hail/annotations/UnsafeOrdering.scala [hunk 1]
```diff
@@ -1,8 +1,15 @@
 package is.hail.annotations
 
+import is.hail.utils._
+import is.hail.asm4s._
+
 abstract class UnsafeOrdering extends Ordering[RegionValue] with Serializable {
   def compare(r1: MemoryBuffer, o1: Long, r2: MemoryBuffer, o2: Long): Int
 
+  def compare(r1: Code[MemoryBuffer], o1: Code[Long], r2: Code[MemoryBuffer], o2: Code[Long]): Code[Int] =
+    // FIXME
```
Comment (danking) [inline src/main/scala/is/hail/annotations/UnsafeOrdering.scala]: oh haha, I never actually implemented this part. I'll do that now, @patrick-schultz, my bad!

**Comment by danking** [comment]
@patrick-schultz Maybe I should make UnsafeOrdering a separate PR...

**Comment by patrick-schultz** [comment]
Yeah, UnsafeOrdering feels like a separate PR. I'll focus on the IR sets for now, so you can give that half to me. You can give the UnsafeOrdering half to me too, but it would still be helpful to me to keep them separate.

**Comment by danking** [comment]
I extracted `UnsafeOrdering` into #2519.

File: src/main/scala/is/hail/expr/ir/Compile.scala [hunk 2]
```diff
@@ -317,8 +317,105 @@ object Compile {
...
+                    srvb.addIRIntermediate(elementType)(velement))),
+                Code(
+                  i := 0,
+                  mx := tArray.isElementMissing(region, s, i),
+                  x := mx.mux(0L, tArray.loadElement(region, s, i)),
+                  Code.whileLoop(i < len && !mx && (ord.compare(region, x, region, elementPointer)(fb, mb) < 0),
+                    mx.mux(srvb.setMissing(), srvb.addRegionValue(elementType)(x)),
+                    mx := tArray.isElementMissing(region, s, i),
+                    x := mx.mux(0L, tArray.loadElement(region, s, i)),
+                    i++),
```
Comment (patrick-schultz) [inline src/main/scala/is/hail/expr/ir/Compile.scala]: I think the i++ needs to come before the two assignments.

File: src/main/scala/is/hail/expr/ir/Compile.scala [hunk 3]
```diff
@@ -317,8 +317,107 @@ object Compile {
...
+                Code(
+                  melement.mux(
+                    srvb.setMissing(),
+                    srvb.addIRIntermediate(elementType)(velement))),
+                Code(
+                  i := 0,
+                  mx := tArray.isElementMissing(region, s, i),
+                  x := mx.mux(0L, tArray.loadElement(region, s, i)),
+                  Code.whileLoop(i < len && !mx && (ord.compare(region, x, region, elementPointer)(fb, mb) < 0),
+                    mx.mux(srvb.setMissing(), srvb.addRegionValue(elementType)(x)),
```
[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/expr/ir/Compile.scala]: Yeah I'm confused about `UnsafeOrdering` honestly. The int32 ordering, for example doesn't check missingness at all. I made a comment to this effect in #2519 .

[the remark this answers is not in the document]
Comment (danking) [reply src/main/scala/is/hail/expr/ir/Compile.scala]: I addressed the `mx` is always false part though.

File: src/main/scala/is/hail/expr/ir/Compile.scala (hunk 2 above)
Comment (danking) [reply src/main/scala/is/hail/expr/ir/Compile.scala]: done.

File: src/main/scala/is/hail/expr/ir/Compile.scala [hunk 4]
```diff
@@ -317,8 +317,100 @@ object Compile {
...
+        val x = fb.newLocal[Long]
+        val storedElement = fb.newLocal()(TypeToTypeInfo(tArray.elementType)).asInstanceOf[LocalRef[Any]]
+        val elementPointer = fb.newLocal[Long]
+        val ord = elementType.unsafeOrdering(true)
+        val sortedArray = Code(
+          doelement,
+          s := coerce[Long](vset),
+          len := TContainer.loadLength(region, s),
+          melement.mux(
+            tArray.isElementMissing(region, s, len - 1).mux(
```
Comment (patrick-schultz) [inline src/main/scala/is/hail/expr/ir/Compile.scala]: This doesn't work in the case `len = 0` does it?

**Comment by danking** [comment]
Yeah I saw the missing matches and have them fixed locally. I'll push them up soon.

**Comment by danking** [comment]
This needs to be recreated once #2519 (on which this depends) is adapted to handle GenomeReference.

**Pull request closed without merging**